docs/clients/dotnet/21.2/migration-to-gRPC.md

```diff
@@ -0,0 +1,630 @@
+            uncommittedEvents.ToArray(),
+            cancellationToken: cancellationToken
+        );
+
+    public void Append(params EventData[] eventData) =>
+        uncommittedEvents.AddRange(eventData);
+}
+```
+
+
+## Reading Events
+
+### Read direction
+
+The TCP Client have dedicated methods for reading events in the specific direction: 
+- `ReadStreamEventsForwardAsync`,
+- `ReadStreamEventsBackwardAsync`,
+- `ReadAllEventsForwardAsync`,
+- `ReadStreamEventsBackwardAsync`.
+
+In the gRPC client, we unified those methods into methods with the direction parameter:
+- `ReadStreamAsync`,
+- `ReadAllAsync`.
+
+To read a stream forwards, use:
+
+```csharp
+await using var readResult = grpcClient.ReadStreamAsync(
+    Direction.Forwards,
+    streamName,
+    StreamPosition.Start
+);
+```
+
+To read a stream backwards, use: 
+
+```csharp
+await using var readResult = grpcClient.ReadStreamAsync(
+    Direction.Backwards,
+    streamName,
+    StreamPosition.End
+);
+```
+
+Accordingly, to read the `$all` stream forwards, use:
+
+```csharp
+await using var readResult = grpcClient.ReadAllAsync(
+    Direction.Forwards,
+    streamName,
+    Position.Start
+);
+```
+
+and to read the `$all` stream backwards:
+
+```csharp
+await using var readResult = grpcClient.ReadAllAsync(
+    Direction.Backwards,
+    streamName,
+    Position.End
+);
+```
+
+### Read positions
+
+Both TCP and gRPC clients allow reading the stream from a specific position (representing the location of the particular event in the stream). We decided to make positions more explicit accordingly to the stream revision changes in [appending new events](#appending-events). 
+
+We expanded a `StreamPosition` class that centralises stream position handling. It still has a `Start` and an `End` constant representing the logical stream's position. However, they return no longer `long` values but `StreamPosition` class instance. We replaced `long` values with `ulong` as stream position is always positive.
+
+`StreamPosition` class has an overloaded operator for `ulong` value assignment. 
+
+```csharp
+StreamPosition streamPosition = 100L;
+```
+
+You can also convert it from the `StreamRevision`:
+
+```csharp
+var streamPosition = StreamPosition.FromStreamRevision(streamRevision);
+```
+
+For reading from `$all`, the TCP Client has already the `Position` class. gRPC client also has it, but using `ulong` instead of `long` for commit and prepare positions.
+
+### Read result
+
+TCP Client requires paging through the results. You must provide the maximum number of events you want to read in the single read call. In the gRPC client, this is optional. By default, it will try to read all events. To make it efficient, read methods return [IAsyncEnumerable](https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8). Which means that it won't load the whole stream at once but iterate sequentially, reducing the memory pressure.
```

Comment:
```suggestion
The TCP Client requires paging through the results. You must provide the maximum number of events you want to read in the single read call. In the gRPC client, this is optional. By default, it will try to read all events. To make it efficient, read methods return [IAsyncEnumerable](https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8). Which means that it won't load the whole stream at once but iterate sequentially, reducing the memory pressure.
```